### ISSUE TITLE
[sklearn#34068] Callbacks x Metadata-routing

### ISSUE THREAD

**Issue by godofecht** [flooooooooooow/flow#359]
**Mirrored from [scikit-learn#34068](https://github.com/scikit-learn/scikit-learn/issues/34068)**

Original title: Callbacks x Metadata-routing
Original labels: Metadata Routing, Callbacks
Created: 2026-05-20T14:45:47Z

---

> [!WARNING]
> This issue is not yet ready for a PR. If you are interested in contributing to scikit-learn, please have a look at our [contributing guidelines](https://scikit-learn.org/dev/developers/contributing.html), and in particular the sections for [new contributors](https://scikit-learn.org/dev/developers/contributing.html#new-contributors) and the ["Needs triage"](https://scikit-learn.org/dev/developers/contributing.html#issues-tagged-needs-triage) label.

There are 2 (related) things to figure out to enable callbacks to be compatible with metadata-routing. I think it's the first thing to solve before starting to implement callback support in more estimators.

- Do we want to allow passing arbitrary metadata to the callbacks ?
  ```py
  my_business_metric.set_score_request(amount=True)
  callback = ScoringMonitor(scoring=my_business_metric)
  LogisticRegression().set_callbacks(callback).fit(X, y, amount=amount)
  ```
  I'm not sure that it's very interesting for the ScoringMonitor. Do we think that anyone is interested in the evolution of a business metric during fit ?
  Same question for early stopping. Is it something that people do ?
  That being said, even if not useful for these callbacks, that may be useful for custom callbacks that I haven't thought about.

  What does that require ?
  Basically make every estimator a ``router``. And callbacks as well.
  That's not a major difficulty but it means that implementing callback support is tied to metadata-routing if you want full capacity (it might be possible to make it optional and then callbacks would be limited when there's no metadata-routing, as it's the case now)

- How to handle ``X_val`` ?
  Assuming that we do the above point, we're facing a major issue with how ``X_val`` is treated right now, that is as regular metadata. Consider the following:
  ```py
  my_business_metric.set_score_request(amount=True)
  callback = ScoringMonitor(scoring=my_business_metric, eval_on="val")
  LogisticRegression().set_callbacks(callback).fit(X, y, X_val=X_val, y_val=y_val, amount_train=amount, amount_val=amount_val)
  ```
  LogisticRegression is not a consumer of ``amount``: it has no idea which one is for which set. In fact amount_train is not even useful here so I could have just passed amount=amount_val.

  X_val and y_val themselves could arguably be consumed by the estimator (at least there are estimators that transform X so X_val should follow).

  Another thing is: if we request ``amount`` on the scorer then how do we sort out how ``amount_train`` and ``amount_val`` are forwarded to the scorer ?

  All of it comes to me from the fact that ``X_val`` is considered regular metadata while it's not really metadata, it's more a secondary dataset with possibly associated metadata. I'm under the impression that the solution lies in the metadata-routing framework rather than in the callbacks, by differentiating from the start metadata associated the train set and metadata associated to the valida

...(truncated)

---

## Flow context

This issue is mirrored from scikit-learn to explore whether Flow encounters the same problem or can approach it differently. Add analysis below.

**Comment by godofecht** [comment]
Moved to [56](https://github.com/godofecht/flow-scikit/issues/56) in the flow-scikit repository. Further library-specific triage and implementation will happen there.

**Comment by godofecht** [comment]
Moved to [57](https://github.com/godofecht/flow-scikit/issues/57) in the flow-scikit repository. Further library-specific triage and implementation will happen there.
